adapters/tlsio_ssl_esp8266.c

```diff
@@ -0,0 +1,960 @@
+            break;
+
+        case TLSIO_STATE_CLOSING:
+            tls_io_instance->tlsio_state = TLSIO_STATE_NOT_OPEN;
+
+            if (tls_io_instance->on_io_close_complete != NULL)
+            {
+                tls_io_instance->on_io_close_complete(tls_io_instance->on_io_close_complete_context);
+            }
+            break;
+    }
+}
+
+static void on_underlying_io_open_complete(void* context, IO_OPEN_RESULT open_result)
+{
+    TLS_IO_INSTANCE* tls_io_instance = (TLS_IO_INSTANCE*)context;
+
+    if (tls_io_instance->tlsio_state == TLSIO_STATE_OPENING_UNDERLYING_IO)
+    {
+        if (open_result == IO_OPEN_OK)
+        {
+            tls_io_instance->tlsio_state = TLSIO_STATE_IN_HANDSHAKE;
+
+            if (send_handshake_bytes(tls_io_instance) != 0)
+            {
+                if (xio_close(tls_io_instance->underlying_io, on_underlying_io_close_complete, tls_io_instance) != 0)
+                {
+                    indicate_open_complete(tls_io_instance, IO_OPEN_ERROR);
+                    LogError("Error in xio_close.");
+                }
+            }
+        }
+        else
+        {
+            indicate_open_complete(tls_io_instance, IO_OPEN_ERROR);
+            LogError("Invalid tlsio_state. Expected state is TLSIO_STATE_OPENING_UNDERLYING_IO.");
+        }
+    }
+}
+
+static void on_underlying_io_error(void* context)
+{
+    TLS_IO_INSTANCE* tls_io_instance = (TLS_IO_INSTANCE*)context;
+
+    switch (tls_io_instance->tlsio_state)
+    {
+        default:
+            break;
+
+        case TLSIO_STATE_OPENING_UNDERLYING_IO:
+        case TLSIO_STATE_IN_HANDSHAKE:
+            tls_io_instance->tlsio_state = TLSIO_STATE_NOT_OPEN;
+            indicate_open_complete(tls_io_instance, IO_OPEN_ERROR);
+            break;
+
+        case TLSIO_STATE_OPEN:
+            indicate_error(tls_io_instance);
+            break;
+    }
+}
+
+static int decode_ssl_received_bytes(TLS_IO_INSTANCE* tls_io_instance)
+{
+    int result = 0;
+    unsigned char buffer[64];
+
+    int rcv_bytes = 1;
+    rcv_bytes = SSL_read(tls_io_instance->ssl, buffer, sizeof(buffer));
+    os_printf("ssl recv bytes: %d\n", rcv_bytes);
+    if (rcv_bytes > 0)
+    {
+        if (tls_io_instance->on_bytes_received == NULL)
+        {
+            LogError("NULL on_bytes_received.");
+        }
+        else
+        {
+            tls_io_instance->on_bytes_received(tls_io_instance->on_bytes_received_context, buffer, rcv_bytes);
+        }
+    }
+
+    return result;
+}
+
+static void destroy_openssl_instance(TLS_IO_INSTANCE* tls_io_instance)
+{
+    if (tls_io_instance != NULL)
+    {
+        SSL_free(tls_io_instance->ssl);
+        SSL_CTX_free(tls_io_instance->ssl_context);
+    }
+}
+
+static int add_certificate_to_store(TLS_IO_INSTANCE* tls_io_instance, const char* certValue)
+{
+    return 0;
+}
+
+
+static int create_openssl_instance(TLS_IO_INSTANCE* tlsInstance)
+{
+    return 0;
+}
+
+int tlsio_openssl_init(void)
```

Comment:
I have concerns about a function that do not belongs to the standard interface. It can clearly became a problem in the future. I prefer to joint the init functions to the create and the deinit to destroy. If it is not possible, we should add this function to the xio interface.